Add XDG Base Directory support

4e554c4c opened the pull request:
Test plan:
---------
```
> mkdir ~/.agda
> cabal run agda -- --print-agda-app-dir
/home/pounce/.agda

> rmdir ~/.agda
> cabal run agda -- --print-agda-app-dir
/home/pounce/.config/agda

> AGDA_DIR=/tmp cabal run agda -- --print-agda-app-dir
/tmp

> cabal run agda -- --print-agda-data-dir
/home/pounce/programming/agda/./src/data
```

fixes #6852

4e554c4c committed:
Add XDG Base Directory support

Test plan:
---------
```
> mkdir ~/.agda
> cabal run agda -- --print-agda-app-dir
/home/pounce/.agda

> rmdir ~/.agda
> cabal run agda -- --print-agda-app-dir
/home/pounce/.config/agda

> AGDA_DIR=/tmp cabal run agda -- --print-agda-app-dir
/tmp

> cabal run agda -- --print-agda-data-dir
/home/pounce/programming/agda/./src/data
```

fixes #6852

4e554c4c committed:
Add documentation for --print-agda-app-dir

This commit also changes the documentation for --print-agda-dir,
renaming it to --print-agda-data-dir. It also removes inconsistancies in
naming between $AGDA_DIR and the agda data directory.

4e554c4c commented:
@wenkokke I think the renaming from `--print-agda-dir` to `--print-agda-data-dir` will have an impact on setup-agda
I'm happy to add it back for compatibility reasons, if need-be.

wenkokke commented:
Thanks for the heads up!

I don't think it will, because I manually set the data directory through the environment, as the data directory set by cabal isn't reliable.

wenkokke commented:
With the proposed change, would Agda continue to respond to the `Agdadata_dir` environment variable?

4e554c4c commented:
it should, i didn't touch that at all (only app dir resolution)

4e554c4c commented:
@wenkokke 
ok this was a little bit hard to test, since `cabal run` messes with `Agda_datadir`, but it seems that the environment variable still works.
```
$ Agda_datadir=/tmp dist-newstyle/build/*/ghc-*/Agda-*/build/agda/agda --print-agda-data-dir
/tmp
```

wenkokke commented:
Honestly, it may be easier to add code for resolving finding the Agda app and data directories based on configuration and environment variables, and work on phasing out the use of the Cabal-generated Paths module, given how much trouble it gives us.

plt-amy commented on doc/user-manual/tools/command-line-options.rst:
```diff
@@ -111,13 +111,20 @@ General options
-     of the directory structure holding Agda's data files
-     such as core libraries, style files for the backends etc.
+     Outputs the root of the directory structure holding Agda's data
+     files such as core libraries, style files for the backends etc.
```
```suggestion
     files such as core libraries, style files for the backends, etc.
```

plt-amy replied:
Can you also mention the old command?

4e554c4c replied:
not an oxford comma enjoyer, but it appears to be used fairly universally in this file. added.

plt-amy commented on src/full/Agda/Interaction/Options/Base.hs:
```diff
@@ -1434,7 +1439,10 @@ standardOptions =
                     ]
 
-    , Option []     ["print-agda-dir"] (NoArg printAgdaDirFlag)
+    , Option []     ["print-agda-data-dir"] (NoArg printAgdaDataDirFlag)
```
Since we're quite late in the release cycle for 2.6.4, I'd like to avoid introducing any breaking changes. Can you keep `--print-agda-dir`? Adding a less ambiguous phrasing is fine by me.

4e554c4c replied:
```
> cabal run agda -- --print-agda-data-dir
/home/pounce/programming/agda/./src/data

> cabal run agda -- --print-agda-dir
/home/pounce/programming/agda/./src/data
```

plt-amy reviewed (changes requested):
Very welcome changes, thank you! For the sake of backwards compatibility, I don't think we should get rid of `--print-agda-dir` just yet.

4e554c4c committed:
re-add `--print-agda-dir` option for compatibility

Test plan:
---------
```
> cabal run agda -- --print-agda-data-dir
/home/pounce/programming/agda/./src/data

> cabal run agda -- --print-agda-dir
/home/pounce/programming/agda/./src/data
```

4e554c4c committed:
add oxford comma

plt-amy commented:
@andreasabel Do you think this needs a test case? If not, I'm merging it. (If so, how?)

4e554c4c commented:
Is this going to make it into the 2.6.4 release?

edit: no

jespercockx commented:
@plt-amy Is now a good time to merge this?

plt-amy commented:
I think so, yeah!

andreasabel commented on doc/user-manual/tools/command-line-options.rst:
```diff
@@ -111,13 +111,26 @@ General options
 
+.. option:: --print-agda-app-dir
+
+     .. versionadded:: 2.6.4
```
This needs bumping to 2.6.4.1

andreasabel commented on doc/user-manual/tools/command-line-options.rst:
```diff
@@ -111,13 +111,26 @@ General options
+
+.. option:: --print-agda-data-dir
+
+     .. versionadded:: 2.6.4
```
Ditto

andreasabel commented on doc/user-manual/tools/generating-html.rst:
```diff
@@ -18,10 +18,10 @@ your own CSS file instead of the :download:`default, included one
-  :samp:`cat $(agda --print-agda-dir)/html/Agda.css`.
+  :file:`{${AGDA_DATA_DIR}}/html/Agda.css`.  Since version 2.6.2, the
+  Agda data directory is printed by option
+  :option:`--print-agda-data-dir`. Thus, you can get hold of the CSS
```
Not entirely correct.
From 2.6.2, `--print-agda-dir`, since 2.6.4.1, `--print-agda-data-dir`.

andreasabel commented on doc/user-manual/tools/generating-latex.rst:
```diff
@@ -32,10 +32,10 @@ from :file:`agda.sty`.
+  :file:`{${AGDA_DIR}}/latex/agda.  Since version 2.6.2, the
+  Agda data directory is printed by option
+  :option:`--print-agda-data-dir`. Thus, you can get hold of the class
+  file via :samp:`cat $(agda --print-agda-data-dir)/latex/agda.sty`.
```
Ditto.

andreasabel commented on doc/user-manual/tools/generating-latex.rst:
```diff
@@ -32,10 +32,10 @@ from :file:`agda.sty`.
-  :envvar:`AGDA_DIR` is printed by option :option:`--print-agda-dir`.
-  Thus, you can get hold of the CSS file via
-  :samp:`cat $(agda --print-agda-dir)/latex/agda.sty`.
+  :file:`{${AGDA_DIR}}/latex/agda.  Since version 2.6.2, the
```
This should also be `Agda_datadir`, shouldn't it?

andreasabel commented on doc/user-manual/tools/command-line-options.rst:
```diff
@@ -111,13 +111,26 @@ General options
+     .. versionadded:: 2.6.4
+
+     Outputs the root of the directory structure holding Agda's data
+     files such as core libraries, style files for the backends, etc.
```
Should this mention the env var `Agda_datadir`?

andreasabel commented on doc/user-manual/tools/generating-html.rst:
```diff
@@ -18,10 +18,10 @@ your own CSS file instead of the :download:`default, included one
-  :envvar:`AGDA_DIR` is printed by option :option:`--print-agda-dir`.
-  Thus, you can get hold of the CSS file via
-  :samp:`cat $(agda --print-agda-dir)/html/Agda.css`.
+  :file:`{${AGDA_DATA_DIR}}/html/Agda.css`.  Since version 2.6.2, the
```
I think the old docs `AGDA_DIR` were wrong.  The correct thing (now and then) should be `Agda_datadir`.

andreasabel commented on doc/user-manual/tools/package-system.rst:
```diff
@@ -135,9 +135,13 @@ To be found by Agda a library file has to be listed (with its full path) in a
-overridden by setting the ``AGDA_DIR`` environment variable.
+overridden by setting the :envvar:``AGDA_DIR`` environment variable.
+
+The ``AGDA_DIR`` will fall-back to ``~/.agda``, if it exists, for
```
Use `:envvar:` here?

andreasabel commented on doc/user-manual/tools/package-system.rst:
```diff
@@ -135,9 +135,13 @@ To be found by Agda a library file has to be listed (with its full path) in a
+
+The ``AGDA_DIR`` will fall-back to ``~/.agda``, if it exists, for
+backward compatibility reasons. You can find the precise location of
+``AGDA_DIR``  by running ``agda --print-agda-app-dir``.
```
Ditto.

andreasabel reviewed (commented):
Thanks for this PR.

There seem to be some errors in the docs, partially old errors, partially introduced by this PR.
Also, a CHANGELOG entry is missing so far.

andreasabel commented:
@plt-amy agreed to write the changelog entry.
Amelia, could you also fix the docs (see my comments above)?
